contracts/truefi2/AaveBaseRateOracle.sol

```diff
@@ -0,0 +1,163 @@
+            uint256[BUFFER_SIZE] memory,
+            uint256[BUFFER_SIZE] memory,
+            uint16
+        )
+    {
+        return (totalsBuffer.runningTotals, totalsBuffer.timestamps, totalsBuffer.currIndex);
+    }
+
+    function getAaveVariableBorrowAPY(address _asset) public view returns (uint256) {
+        (, , , , uint128 currentVariableBorrowRate, , , , , , , ) = aavePool.getReserveData(_asset);
+        return uint256(currentVariableBorrowRate).div(1e23);
+    }
+
+    /**
+     * @dev Update the totalsBuffer:
+     * Gets current deposit apy from aave and writes down
+     * new total running value.
+     * If the buffer is filled overwrites the oldest value
+     * with a new one and updates its timestamp.
+     */
+    function update() public offCooldown {
+        uint16 _currIndex = totalsBuffer.currIndex;
+        uint16 nextIndex = (_currIndex + 1) % bufferSize();
+        uint256 apy = getAaveVariableBorrowAPY(asset);
+        uint256 dt = block.timestamp.sub(totalsBuffer.timestamps[_currIndex]);
+        totalsBuffer.runningTotals[nextIndex] = totalsBuffer.runningTotals[_currIndex].add(
+            apy.add(totalsBuffer.lastValue).mul(dt).div(2)
+        );
+        totalsBuffer.timestamps[nextIndex] = block.timestamp;
+        totalsBuffer.lastValue = apy;
+        totalsBuffer.currIndex = nextIndex;
+    }
+
+    /**
+     * @dev Average apy is calculated by taking
+     * the time-weighted average of the aave deposit apys.
+     * Essentially formula given below is used:
+     *
+     *           (v + v_{n-1}) / 2 * (t - t_{n-1}) + sum_{i=1}^{n - 1} (v_i + v_{i-1}) / 2 * (t_i - t_{i-1})
+     * avgAPY = -------------------------------------------------------------------------------------------
+     *                                                  (t - t_0)
+     *
+     * where v_i, t_i are values of the prices and their respective timestamps
+     * stored in the historical buffer, v is a value of current price ant t is its timestamp.
+     * Index n-1 corresponds to the most recent values and index 0 to the oldest ones.
+     *
+     * To avoid costly computations in a loop an optimization is used:
+     * Instead of directly storing aave deposit apys we store calculated numerators from formula above.
+     * This gives us most of the job done for every calculation.
+     *
+     * Notice that whether we are going to use the whole buffer or not
+     * depends on if it is filled up and the value of timeToCover parameter.
+     * @param timeToCover For how much time average should be calculated.
+     * @return Average apy.
+     */
+    function calculateAverageAPY(uint256 timeToCover) public view returns (uint256) {
+        require(
+            1 days <= timeToCover && timeToCover <= 365 days,
+            "AaveBaseRateOracle: Expected amount of time in range 1 to 365 days"
+        );
+        // estimate how much buffer we need to use
+        uint16 bufferSizeNeeded = uint16(timeToCover.div(cooldownTime));
```

Comment:
I'm a little concerned about what may happen, if `update` gets called with a one day delay, because of any random reason. Wouldn't we then need one more, update than we actually have stored? And wouldn't then the starting point be 0 (if we haven't updated the whole buffer yet)?

I think I would feel much more secure, if we passed `numberOfUpdates` instead of `timeToCover` to this function.

Reply:
Thanks, I like this idea of using `numberOfUpdates` instead of `timeToCover`. I'm going to implement it.